src/main/java/org/apache/commons/compress/archivers/tar/TarFile.java

```diff
@@ -0,0 +1,712 @@
+/*
+ *  Licensed to the Apache Software Foundation (ASF) under one or more
+ *  contributor license agreements.  See the NOTICE file distributed with
+ *  this work for additional information regarding copyright ownership.
+ *  The ASF licenses this file to You under the Apache License, Version 2.0
+ *  (the "License"); you may not use this file except in compliance with
+ *  the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ */
+package org.apache.commons.compress.archivers.tar;
+
+import java.io.ByteArrayOutputStream;
+import java.io.Closeable;
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.ByteBuffer;
+import java.nio.channels.SeekableByteChannel;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+
+import org.apache.commons.compress.archivers.zip.ZipEncoding;
+import org.apache.commons.compress.archivers.zip.ZipEncodingHelper;
+import org.apache.commons.compress.utils.ArchiveUtils;
+import org.apache.commons.compress.utils.BoundedInputStream;
+import org.apache.commons.compress.utils.BoundedNIOInputStream;
+import org.apache.commons.compress.utils.BoundedSeekableByteChannelInputStream;
+import org.apache.commons.compress.utils.SeekableInMemoryByteChannel;
+
+/**
+ * The TarFile provides random access to UNIX to archives.
+ * @since 1.21
+ */
+public class TarFile implements Closeable {
```

Comment:
Many methods in `TarFile` looks pretty like the ones in `TarArchiveInputStream`(e.g. `buildSparseInputStreams`, `getLongNameData`,  `buildSparseInputStreams`). Maybe we can find some way to share these methods?

Reply:
The duplication is really on purpose from my side in the current state of the PR. I didn't wanted to limit myself to early from making the needed changes to get this work. Also I wanted to limit changes to existing code to avoid breaking it. Sharing some of the code can be quite a challenge because the data source is so different.
`buildSparseInputStream`: The difference here is that the sparse streams are saved differently in the TarFile than in the Stream. In the TarFile the streams need to be saved on a per entry basis to allow random access. I think this could also be done in the stream but for the first version I didn't wanted to change too much existing code.
`getLongNameData`: The only difference is the source from where the data is read (InputStream vs. SeekableByteChannel). A problem could be the call to `getNext(Tar)Entry` which has a lot of sideeffects.